### PR TITLE
[ControlFlowGraph] Initial

### PR DESCRIPTION
**Pull request by ovr**
Hey!

To see what changed...

Thanks

### CODE REVIEW

**Comment by ovr** [comment]
@ddmler Can you take a look?

**Comment by K-Phoen** [comment]
This looks great! 👍 

I'll try to review it soon (in the next few days) :)

**Comment by ovr** [comment]
For now It support only for functions, it's needed to tests when I am working on it
Soon this will be on all `Definitions`

Next things will be

- [ ] SSA + Phi
- [ ] Rewrite Visitor and Printers
- [ ] Unreacheble Analyzer
- [ ] Support more nodes
- [ ] Simple infinity loop analyzer

File: src/ControlFlow/Block.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+     */
+    public function setExit(Block $exit)
+    {
+        $this->exit = $exit;
+    }
+
+    /**
+     * @return AbstractNode[]
+     */
+    public function getChildrens()
```
Comment (ddmler) [inline src/ControlFlow/Block.php]: Children is the correct name not Childrens

File: src/ControlFlow/Block.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+namespace PHPSA\ControlFlow;
+
+use PHPSA\ControlFlow\Node\AbstractNode;
+
+class Block
+{
+    /**
+     * @var AbstractNode[]
+     */
+    protected $childrens = [];
```
Comment (ddmler) [inline src/ControlFlow/Block.php]: "children"

File: src/ControlFlow/BlockTraverser.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    public function traverse(ControlFlowGraph $cfg)
+    {
+        $rootBlock = $cfg->getRoot();
+
+        /** @var AbstractVisitor $visitor */
+        foreach ($this->visitors as $visitor) {
+            $visitor->enterBlock($rootBlock);
+        }
+
+        $childrens = $rootBlock->getChildrens();
```
Comment (ddmler) [inline src/ControlFlow/BlockTraverser.php]: "children"

File: src/ControlFlow/ControlFlowGraph.php [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+     */
+    protected $lastBlockId = 1;
+
+    /**
+     * @var Block
+     */
+    protected $root;
+
+    /**
+     * @var array
```
Comment (ddmler) [inline src/ControlFlow/ControlFlowGraph.php]: var Block[] instead of array?

File: src/ControlFlow/ControlFlowGraph.php [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+        if ($statement instanceof Function_) {
+            if ($statement->stmts) {
+                $this->passNodes($statement->stmts, $this->root);
+            }
+        }
+    }
+
+    /**
+     * @param array $nodes
```
Comment (ddmler) [inline src/ControlFlow/ControlFlowGraph.php]: @param \PhpParser\Node[] $nodes

File: src/ControlFlow/Node/AbstractNode.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * @author Patsura Dmitry https://github.com/ovr <talk@dmtry.me>
+ */
+
+namespace PHPSA\ControlFlow\Node;
+
+abstract class AbstractNode
+{
+    /**
+     * Can this node exit?
```
Comment (ddmler) [inline src/ControlFlow/Node/AbstractNode.php]: I suggest to change comment to "will this node exit"

File: src/ControlFlow/Node/Expr/BinaryOp/Equal.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
+<?php
+/**
+ * @author Patsura Dmitry https://github.com/ovr <talk@dmtry.me>
+ */
+
+namespace PHPSA\ControlFlow\Node\Expr\BinaryOp;
+
+use PHPSA\ControlFlow\Node\AbstractNode;
+
+class Equal extends AbstractNode
```
Comment (ddmler) [inline src/ControlFlow/Node/Expr/BinaryOp/Equal.php]: maybe extend a new abstract node called "BinaryOpNode" that has $left, $right, $result and getSubVariables?
It's the same for all binary ops it seems or is this gonna change later?

File: src/ControlFlow/Block.php (hunk 2 above)
Comment (ovr) [inline src/ControlFlow/Block.php]: Thanks

File: src/ControlFlow/Block.php (hunk 1 above)
Comment (ovr) [inline src/ControlFlow/Block.php]: Thanks

File: src/ControlFlow/BlockTraverser.php (hunk 3 above)
Comment (ovr) [inline src/ControlFlow/BlockTraverser.php]: Thanks

File: src/ControlFlow/ControlFlowGraph.php (hunk 4 above)
Comment (ovr) [inline src/ControlFlow/ControlFlowGraph.php]: Thanks

File: src/ControlFlow/Node/AbstractNode.php (hunk 6 above)
Comment (ovr) [inline src/ControlFlow/Node/AbstractNode.php]: Thanks

File: src/ControlFlow/Node/Expr/BinaryOp/Equal.php (hunk 7 above)
Comment (ovr) [inline src/ControlFlow/Node/Expr/BinaryOp/Equal.php]: Yeap, sure, was thinking to fix it later

File: src/ControlFlow/ControlFlowGraph.php [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+     * If current block is not empty, lets create a new one
+     *
+     * @param Block $block
+     * @return Block
+     */
+    protected function createNewBlockIfNeeded(Block $block)
+    {
+        if ($block->getChildren()) {
+            $block->setExit(
+                $block = new Block($this->lastBlockId++)
```
Comment (algo13) [inline src/ControlFlow/ControlFlowGraph.php]: [Assignment in sub-expressions](https://sonarqube.com/coding_rules#rule_key=c%3AAssignmentInSubExpression)

This seems not to be even circular reference, is it intentional ?

Maybe the same as

```php
    protected function createNewBlockIfNeeded(Block $block)
    {
        if ($block->getChildrens()) {
            $block = new Block($this->lastBlockId++);
        }

        return $block;
    }
```

Comment (ovr) [inline src/ControlFlow/ControlFlowGraph.php]: `setExit` is expected, previous block should exit (goto/jump) to new one, without it We will not reach new block

Examples of building new block

```php
function test() {
    $a = 1;

    test:
    $b = 1;

    goto test;
}
```

image

as you can I use `label` to test `createNewBlockIfNeeded`
`Block#1` will exit to `Block#2`

Comment (algo13) [inline src/ControlFlow/ControlFlowGraph.php]: Okay, I understand.
I got confused by 'Assignment in sub-expressions'.

same as:

```php
    protected function createNewBlockIfNeeded(Block $block)
    {
        if ($block->getChildrens()) {
            $blockNew = new Block($this->lastBlockId++);
            $block->setExit($blockNew);
            return $blockNew;
        }

        return $block;
    }
```

**Pull request merged**